hbase-server/src/main/java/org/apache/hadoop/hbase/master/MobFileCompactionChore.java

```diff
@@ -0,0 +1,224 @@
+                admin.majorCompact(htd.getTableName(), hcd.getName());
+              } else {
+                LOG.debug("Performing compaction in batches, {}={}",
+                  "hbase.mob.compaction.batch.size", regionBatchSize);
+                performMajorCompactionInBatches(admin, htd, hcd);
+              }
+            } else {
+              LOG.debug("Skipping column family {} because it is not MOB-enabled",
+                hcd.getNameAsString());
+            }
+          } catch (IOException e) {
+            LOG.error(
+              "Failed to compact table=" + htd.getTableName() + " cf=" + hcd.getNameAsString(), e);
+          } catch (InterruptedException ee) {
+            Thread.currentThread().interrupt();
+            master.reportMobCompactionEnd(htd.getTableName());
+            LOG.warn(
+              "Failed to compact table=" + htd.getTableName() + " cf=" + hcd.getNameAsString(), ee);
+            // Quit the chore
+            return;
+          }
+        }
+        if (reported) {
+          master.reportMobCompactionEnd(htd.getTableName());
+          reported = false;
+        }
+      }
+    } catch (IOException e) {
+      LOG.error("Failed to compact", e);
+    }
+  }
+
+  @VisibleForTesting
+  public void performMajorCompactionInBatches(Admin admin, TableDescriptor htd,
+      ColumnFamilyDescriptor hcd) throws IOException, InterruptedException {
+
+    List<RegionInfo> regions = admin.getRegions(htd.getTableName());
+    if (regions.size() <= this.regionBatchSize) {
+      LOG.debug("Performing compaction in non-batched mode, regions={}, batch size={}",
+        regions.size(), regionBatchSize);
+      admin.majorCompact(htd.getTableName(), hcd.getName());
+      return;
+    }
+    // Shuffle list of regions in case if they come ordered by region server
+    Collections.shuffle(regions);
+    // Create first batch
+    List<RegionInfo> toCompact = new ArrayList<RegionInfo>();
+    for (int i = 0; i < this.regionBatchSize; i++) {
+      toCompact.add(regions.remove(0));
+    }
+
+    // Start compaction now
+    for (RegionInfo ri : toCompact) {
+      startCompaction(admin, htd.getTableName(), ri, hcd.getName());
+    }
+
+    List<RegionInfo> compacted = new ArrayList<RegionInfo>();
+    int totalCompacted = 0;
+    while (!toCompact.isEmpty()) {
+      // Check status of active compactions
+      for (RegionInfo ri : toCompact) {
+        try {
+          if (admin.getCompactionStateForRegion(ri.getRegionName()) == CompactionState.NONE) {
+            totalCompacted++;
+            LOG.info("Finished major compaction: table={} region={}, compacted regions={}",
+              htd.getTableName(), ri.getRegionNameAsString(), totalCompacted);
+            compacted.add(ri);
+          }
+        } catch (IOException e) {
+          LOG.warn("Could not get compaction state for region {}", ri.getEncodedName());
```

Comment:
should we eventually pull the region out of the `toCompact` list if it keeps failing? otherwise don't we risk getting stuck here and needing a master restart to get any other tables compacted?

Reply:
Yes, I updated logic to remove failed regions from toCompact list to avoid endless loop
in case if some region fails due to network or region server issues. They will be compacted 
in a next chore run.